source/MaterialXCore/Types.h

```diff
@@ -173,6 +174,32 @@ template <class V, class S, size_t N> class VectorN : public VectorBase
         return *this;
     }
 
+    /// @}
+    /// @name Geometric operators
+    /// @{
+
+    /// Magnitude of a vector
+    S magnitude() const
+    {
+        S result = (S)0;
+        for (size_t i = 0; i < N; i++)
+        {
+            result += _arr[i] * _arr[i];
+        }
+        result = std::sqrt(result);
+        return result;
+    }
+
+    /// Normalization
+    void normalize()
+    {
+        S mag1 = (S)1 / magnitude();
```

Comment:
I would recommend omitting the "multiply by reciprocal" optimization, since it won't be expected by the client, and could exaggerate floating-point errors for vectors with small magnitudes.  Also, we can take advantage of existing vector operators and write this method more compactly:

```
*this = *this / magnitude();
```